File: app/src/main/java/com/jorbital/gymstat/views/RoutinesActivity.java
package com.jorbital.gymstat.views;

import android.content.Intent;
import android.os.Bundle;
import android.support.constraint.ConstraintLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import com.jorbital.gymstat.R;
import com.jorbital.gymstat.utils.BaseActivityWithNavDrawer;
import com.jorbital.gymstat.viewmodels.RoutinesViewModel;

public class RoutinesActivity extends BaseActivityWithNavDrawer
{
    private RoutinesViewModel vm;

    RecyclerView routinesRv;
    ConstraintLayout noRoutinesLayout;

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);

        if (vm.getAllRoutines().isEmpty())
            noRoutinesLayout.setVisibility(View.VISIBLE);

        routinesRv.setHasFixedSize(true);

        routinesRv.setLayoutManager(new LinearLayoutManager(this));
    }

    @Override
    protected void setLayout()
    {
        setContentView(R.layout.activity_routines);
    }

    @Override
    protected void createViewModel()
    {
        vm = new RoutinesViewModel(realm);
        vm.makeListOfRoutines();
    }

    @Override
    protected void updateViewFromViewModel()
    {
        if (routinesRv.getAdapter() == null)
            routinesRv.setAdapter(new RoutinesAdapter(vm.getAllRoutines(), true));
        //else
           //update the list
    }

    public void addRoutine(View view)
    {
        Intent intent = new Intent(this, EditRoutineActivity.class);
        startActivity(intent);
    }

    public void startFreeWorkout(View view)
    {
        Intent intent = new Intent(this, WorkoutActivity.class);
        startActivity(intent);
    }

    public void plateCalculator(View view)
    {
        new PlateCalcDialog(this).show();
    }
}

File: app/src/main/java/com/jorbital/gymstat/views/RoutinesActivity.kt
package com.jorbital.gymstat.views

import kotlinx.android.synthetic.main.activity_routines.*

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.view.View

import com.jorbital.gymstat.R
import com.jorbital.gymstat.data.Routine
import com.jorbital.gymstat.utils.BaseActivityWithNavDrawer
import com.jorbital.gymstat.viewmodels.RoutinesViewModel
import io.realm.RealmResults

class RoutinesActivity : BaseActivityWithNavDrawer() {
    private var vm: RoutinesViewModel? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        vm = ViewModelProviders.of(this).get(RoutinesViewModel::class.java)

        vm!!.allRoutines?.observe(this, Observer { updateViewFromViewModel() })

        routinesRv.setHasFixedSize(true)
        routinesRv.layoutManager = LinearLayoutManager(this)

        updateViewFromViewModel()
    }

    override fun setLayout() {
        setContentView(R.layout.activity_routines)
    }

    private fun updateViewFromViewModel() {
        if (vm == null)
            return

        if (vm!!.noRoutines()!!) {
            noRoutinesLayout.visibility = View.VISIBLE
            return
        } else
            noRoutinesLayout.visibility = View.GONE

        if (routinesRv.adapter == null)
            routinesRv.adapter = RoutinesAdapter(vm!!.allRoutines?.value, true)
        else {
            //update the list, maybe not needed if the realm stuff works like magic
        }
    }

    fun addRoutine(view: View) {
        val intent = Intent(this, EditRoutineActivity::class.java)
        startActivity(intent)
    }

    fun startFreeWorkout(view: View) {
        val intent = Intent(this, WorkoutActivity::class.java)
        startActivity(intent)
    }

    fun plateCalculator(view: View) {
        PlateCalcDialog(this).show()
    }
}